### PR TITLE
Add partial update support to AttributeManager

### PR DESCRIPTION
**Pull request by Pessimistress**
For #1576 [RFC](https://github.com/uber/deck.gl/blob/master/dev-docs/RFCs/vNext/partial-updates-rfc.md)

This is a non-breaking change -- any custom attribute updater now receive `startIndex` `endIndex` parameters. If it doesn't handle them, it would perform a full update on the attribute array just like before.

#### Change List
- Add optional `startIndex` `endIndex` parameters to `AttributeManager.update`
- Docs
- Unit test

#### TODO
- Add a test app for this feature
- Handle partial updates in core layers (e.g. `PathTesselator` and `PolygonTesselator`)
- Validate the provided range -- in path and polygon use cases, if the old and new vertex counts in the partial range do not match, partial update would corrupt the attribute array. This is less of a concern if we limit the usage to internal scenarios.

### CODE REVIEW

**Comment by Pessimistress** [comment]
@ibgreen @georgios-uber Just calling `attributeManager.update` does not update the attributes, you also need to mark an attribute as `needsUpdate`. I wonder if it makes more sense to trigger a partial update with `attributeManager.invalidate(triggerName, range)` from the layer author's perspective.

File: modules/core/src/lib/attribute.js [hunk 1]
```diff
@@ -184,29 +184,35 @@ export default class Attribute extends BaseAttribute {
     return false;
   }
 
-  updateBuffer({numInstances, bufferLayout, data, props, context}) {
+  updateBuffer({numInstances, bufferLayout, data, startIndex, endIndex, props, context}) {
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: With the "table" nomenclature I have been propagating for when it comes to `data` related props and names, these parameters could be called `firstRow`, `lastRow`.

I think that naming would make these parameter easier to understand at a glance, i.e. is this:
- Index into the original data array (data table)?
- index into the generated vertex arrays?
- Index into the indices array?

Calling them row would make it obvious that this is the first case.

File: modules/core/src/lib/attribute.js [hunk 2]
```diff
@@ -184,29 +184,35 @@ export default class Attribute extends BaseAttribute {
...
         // Full update
         this.update({
           value: this.value,
           constant: this.constant
         });
       } else {
+        const startOffset = Number.isFinite(startIndex) ? this._getVertexOffset(startIndex) : 0;
+        const endOffset = Number.isFinite(endIndex)
+          ? this._getVertexOffset(endIndex)
+          : numInstances * this.size;
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Does this only support instanced data? This wouldn't work for a custom geometry layer?

File: modules/core/src/lib/attribute.js [hunk 3]
```diff
@@ -295,6 +301,12 @@ export default class Attribute extends BaseAttribute {
   }
 
   // PRIVATE HELPER METHODS
+  _getVertexOffset(index, bufferLayout = this.bufferLayout) {
+    if (bufferLayout) {
+      return bufferLayout.reduce((sum, cur, idx) => (idx < index ? sum + cur : sum), 0) * this.size;
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Oh no, a `reduce` statement, and with abbreviated parameter names... ;) 

I just find it much harder to verify at a glance that a `reduce` calculation is correct compared to a simple `for` loop.

Plus it looks like it always iterates to the end even though it doesn't have to?

File: modules/core/src/lib/attribute.js [hunk 4]
```diff
@@ -295,6 +301,12 @@ export default class Attribute extends BaseAttribute {
   }
 
   // PRIVATE HELPER METHODS
+  _getVertexOffset(index, bufferLayout = this.bufferLayout) {
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Do we need the `bufferLayout = this.bufferLayout`? 

I sometimes have used such code myself by I do think it looks a little "iffy", mixing state and input parameters in a somewhat subtle way.

File: modules/core/src/utils/iterable-utils.js [hunk 5]
```diff
@@ -26,7 +26,7 @@ const placeholderArray = [];
  * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
  * and a "context" tracker from the given data
  */
-export function createIterable(data) {
+export function createIterable(data, startIndex = 0, endIndex) {
```
Comment (ibgreen) [inline modules/core/src/utils/iterable-utils.js]: For clarify maybe add a default value to the last param as well `endIndex = Infinty` or `null`...

File: docs/api-reference/attribute-manager.md [hunk 6]
```diff
@@ -132,6 +134,8 @@ Parameters:
 * `buffers` (Object) - pre-allocated buffers
 * `props` (Object) - passed to updaters
 * `context` (Object) - Used as "this" context for updaters
+* `startIndex` (Number) - start index in the data object, if only a subset of data needs to be considered in the update
```
Comment (ibgreen) [inline docs/api-reference/attribute-manager.md]: We discussed supporting multiple ranges. Assuming that translates into multiple calls to this function.

Comment (Pessimistress) [reply docs/api-reference/attribute-manager.md]: I am removing this parameter from `attributeManager.update` for now and implementing multiple range support in a follow up PR.

**Pull request merged**